### PR TITLE
DAOS-6174 control: Adjust pool create retry behavior

### PR DESCRIPTION
**Pull request by mjmac**
Moves the logic for retrying pool creation operations to
the client side of the Control API. In hindsight,
this should have been the approach all along because it will
work better with multiple MS replicas. Makes a change to
the target creation logic to intercept a -DER_EXISTS and
convert it to -DER_AGAIN to signal that a pool creation or
cleanup process is already in progress and that the
control plane should retry later.

Also adds support for request deadlines to better control client
behavior. Bumps default pool create timeout to 10 minutes to
accommodate very large pools. Longer-term, we should set this
timeout based on SCM size when we understand the relationship
better.

### CODE REVIEW

**Comment by mjmac** [comment]
@liw: Added you for comment on the change to per-try pool UUIDs. Overall I think that we should move the retry logic back out to the client to work better with multiple MS replicas, but I'm wondering whether it would be a mistake to make each retry have its own pool UUID. I can see a case being made for one user request being tied to one UUID, in order to avoid accidentally creating multiple pools due to some error in the retry logic.

**Review by tanabarr — DISMISSED** [review]

File: src/control/lib/control/request.go [hunk 1]
```diff
@@ -80,12 +98,45 @@ func (r *request) AddHost(hostAddr string) {
...
+// returns false, indicating that the default request implementation
+// is not retryable.
+func (r *request) canRetry(_ error, _ uint) bool {
+	return false
+}
+
+// onRetry implements the retryer interface and always returns
+// an error. Callers should check the result of canRetry() before
+// calling retry, in order to avoid wasting effort on a request
+// that does not implement its own retry logic.
```
Comment (kjacque) [inline src/control/lib/control/request.go:129]: This part of the comment seems to be more about the interface in general than about this specific implementation. I think it'd be helpful to put any general comments about how the interface aught to be used near the interface definition.

File: src/control/lib/control/request.go [hunk 2]
```diff
@@ -97,14 +148,58 @@ func (r *msRequest) isMSRequest() bool {
...
+	// embeds this type.
+	retryTestFn func(error, uint) bool
+	// retryFun defines a function that will run on every retry iteration.
+	retryFn func(context.Context, uint) error
+}
+
+// retryAfter returns a channel to be waited on in order to pause
+// a goroutine's execution for some duration. For safety, the
+// channel should be waited on in a select with a context in order to
+// avoid blocking a goroutine forever.
```
Comment (kjacque) [inline src/control/lib/control/request.go:168]: Another comment that seems a bit more general, about how callers should behave with the interface.

**Review by kjacque — APPROVED** [review]
LGTM. Just a couple minor comments about comments.

**Comment by liw** [comment]
> ~@liw: Added you for comment on [...] associating all retries with the same pool UUID. This seems cleanest and least likely to result in ambiguous behavior.

Agreed. I've left a comment on DAOS-6174.

File: src/control/lib/control/request.go (hunk 2 above)
Comment (mjmac) [reply src/control/lib/control/request.go:168]: fixed

File: src/control/lib/control/request.go (hunk 1 above)
Comment (mjmac) [reply src/control/lib/control/request.go:129]: fixed

**Comment by mjmac** [comment]
Passed build and unit tests. No need to re-run functional tests as the changes since the last full run were to documentation and to fix a codespell complaint.

**Pull request merged**